### PR TITLE
optimize SLN with large dimension

### PR DESCRIPTION
**Pull request by yufenglee**
### Description

Optimize SkipLayerNorm for large dimension (>=2048) by handling 8 elements in one thread. It avoid the re-writing and re-loading sum of input, skip and bias to main memory. It reduces the latency of dimension 4096 with small batch size from ~18us to ~3.8us on A100.

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu [hunk 1]
```diff
@@ -198,38 +203,42 @@ void LaunchSkipLayerNormKernel(
...
-        LAUNCH_SKIP_LAYER_NORM_KERNEL_SMALL(1);                               \
-      } else {                                                                \
-        constexpr int block_size = 256;                                       \
-        LAUNCH_SKIP_LAYER_NORM_KERNEL();                                      \
-      }                                                                       \
-    }                                                                         \
+#define CASE_NEXT_SIZE(next_size_value)                                         \
+  case next_size_value: {                                                       \
+    static_assert(next_size_value >= kSizes[0] && next_size_value <= kMaxSize); \
+    if constexpr (next_size_value >= 8 * 256) {                                 \
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu:209]: Please also test 768 and 1024.  TRT uses unroll 8 for them. I think we can have a try as well.

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu [hunk 2]
```diff
@@ -198,38 +203,42 @@ void LaunchSkipLayerNormKernel(
...
-    }
+    CASE_NEXT_SIZE(kSizes[0]);
     CASE_NEXT_SIZE(kSizes[1]);
     CASE_NEXT_SIZE(kSizes[2]);
     CASE_NEXT_SIZE(kSizes[3]);
     CASE_NEXT_SIZE(kSizes[4]);
     CASE_NEXT_SIZE(kSizes[5]);
+    CASE_NEXT_SIZE(kSizes[6]);
+    CASE_NEXT_SIZE(kSizes[7]);
     // kMaxSize shall not run vectorized kernel since ld might be larger than kMaxSize.
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu:242]: remove this comment.

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu [hunk 3]
```diff
@@ -51,35 +51,34 @@ half maybe2half(float x) {
 
 // Using only power of 2 numbers will lead to waste of compute for same size such as 768, which is a very common case
 // in BERT. Ideally we can step by wrap_size * num_unroll, but listing too many steps will cause long compile time.
-constexpr int kSizes[] = {32, 64, 128, 384, 768, 1024, 2048};
+constexpr int kSizes[] = {128, 384, 768, 1024, 2048, 4096, 5120, 8192};
 constexpr size_t kNumOfSizes = sizeof(kSizes) / sizeof(kSizes[0]);
 constexpr int kMaxSize = kSizes[kNumOfSizes - 1];
 constexpr int kMinBlockSize = 32;
-constexpr int kMaxBlockSize = 256;
+constexpr int kMaxBlockSize = 1024;
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu:58]: use max_threads_per_block from prop?

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu [hunk 4]
```diff
@@ -51,35 +51,34 @@ half maybe2half(float x) {
 
 // Using only power of 2 numbers will lead to waste of compute for same size such as 768, which is a very common case
 // in BERT. Ideally we can step by wrap_size * num_unroll, but listing too many steps will cause long compile time.
-constexpr int kSizes[] = {32, 64, 128, 384, 768, 1024, 2048};
+constexpr int kSizes[] = {128, 384, 768, 1024, 2048, 4096, 5120, 8192};
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu:54]: will it affect perf of smaller sequence length?

File: onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu (hunk 1 above)
Comment (yufenglee) [reply onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu:209]: the main performance gain comes from that it avoids writing and re-loading sum of input, skip and bias to main memory by unrolling. 768 and 1024 have already done this. There may be gain but will be small. Let's try it out and create a new PR if it does give gain.

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**